Performant STATE message flush

airhorns opened the pull request:
This is on top of #120 and a follow up to #128. 

In order for orchestrating systems to properly keep track of the STATE of Singer streams, target-postgres needs to emit the STATE messages it receives from the tap to stdout for the orchestrator to persist. This keeps target-postgres Singer-spec compliant and lets users pass that STATE back into the tap to incrementally load from any big sources.

The Singer spec however does not specify what exactly is in STATE messages, so, they can relate to any or all of the active streams. That means that target-postgres doesn't and can't know which records that it may have buffered in memory are "covered" by an incoming STATE record. If target-postgres eagerly emitted the STATE record to stdout, but didn't flush all the records, the system becomes open to inconsistency. If the STATE message were persisted by the outside orchestrator, but then the process crashed later, the buffers in target-postgres process would be lost and data dropped. The implementation from #120 suffers from this bug, see #120 for more discussion.

This commit prevents this bug by "late" flushing buffered STATE messages when all the records that arrived prior to that STATE message have been flushed. STATE messages are "delayed" until the records that came before have all been put on a buffered stream and then flushed. target-snowflake implements this here: https://gitlab.com/meltano/target-snowflake/blob/master/target_snowflake/target_snowflake.py.

This is accomplished by keeping a low watermark of the least recently arriving record for each stream, and upon flushing a buffer, checking to see if there are any unflushed STATE messages that have become safe to flush because they are below that watermark. The way to think about this is as a "safety cursor" that lags behind the incoming stream, pointing to the point in the stream where all records before that point have been saved to the database. Some records ahead of that point will likely have been saved, as different streams records may be interleaved or arriving at different rates, but because of the whole "states are on one timeline different than each stream" thing, the STATE message has to wait for all prior records to be flushed.

This is implemented using a new StateTracker class, which wraps a bit of business logic around the old streams dict that is used to hold the BufferedSingerStream objects. I'm happy to get feedback on the style or further refactorings that this might prompt so please review away. I tried to make this change in a not super invasive way such that we could potentially chunk up further changes into future PRs.

awm33 commented:
@airhorns This is great!

I added a few comments inline.

But once thing to think about, since `StateTracker`  is now responsible for handling certain messages as well as the root for streams, perhaps it should really be called `StreamTracker` or `Streams` or `MessageHandler`?

airhorns commented:
@awm33 ok, I will rename it. Also I don't see your comments, is your review maybe pending still? Also, also, do you want me to go further with the refactor right now and move more into the `StreamTracker` or is how it is suitable for you?

AlexanderMann commented:
Hey @airhorns, I'm dedicating my day to all things Target-Postgres and am hoping to get you critical feedback here etc.

awm33 commented:
@airhorns @AlexanderMann Sorry! Forgot to hit Submit Review

AlexanderMann commented on target_postgres/state_tracker.py:
```diff
@@ -0,0 +1,78 @@
+        # If they occurred after some records that haven't yet been flushed, they aren't safe to emit.
+        # Because records arrive at different rates from different streams, we take the earliest unflushed record as the threshold for what
+        # STATE messages are safe to emit.
+        all_flushed_watermark = min(self.stream_flush_watermarks.values(), default=0)
```
If we wanted to get crazy about performance, this'd prolly make sense to be a min heap. I'd personally vote for this to be tackled later if there're tonnes of open streams at any given time etc.

AlexanderMann commented on tests/test_target_tools.py:
```diff
@@ -70,3 +76,174 @@ def test_loading__invalid__records__threshold():
+    assert json.loads(output[0])['test'] == 'state-4'
+
+
+def test_state__doesnt__emit__when__it__isnt__different__than__the__previous__emission(capsys):
```
nit: underscores 👍

AlexanderMann commented on tests/test_target_tools.py:
```diff
@@ -70,3 +76,174 @@ def test_loading__invalid__records__threshold():
+DogStream.schema['stream'] = 'dogs'
+
+
+def test_state__doesnt__emit__when__only__one__of__several__streams__is__flushing(capsys):
```
nit: underscores 👍

AlexanderMann commented on tests/test_target_tools.py:
```diff
@@ -70,3 +76,174 @@ def test_loading__invalid__records__threshold():
+    assert json.loads(output[0])['test'] == 'state-4'
+
+
+def test_state__emits__most__recent__state__when__final__flush__occurs(capsys):
```
nit: underscores 👍

AlexanderMann commented on tests/test_target_tools.py:
```diff
@@ -70,3 +76,174 @@ def test_loading__invalid__records__threshold():
+    assert len(output) == 0
+
+
+def test_state__emits__only__messages__when__all__records__before__have__been__flushed(capsys):
```
nit: underscores 👍

AlexanderMann commented on target_postgres/state_tracker.py:
```diff
@@ -0,0 +1,78 @@
+from collections import deque
+import json
+import sys
```
nit: alphabetize the imports in here 👍

airhorns commented on target_postgres/state_tracker.py:
```diff
@@ -0,0 +1,78 @@
+
+    def handle_state_message(self, value):
+        if self.emit_states:
+            self.state_queue.append((value, self.message_counter))
```
I agree, but I did it below in the `_emit_safe_queued_states` function so that the diff is done as late as possible. That way it's comparing states it would actually emit and skipping any intermediates in the queue that might be different but not matter.

airhorns commented on tests/test_target_tools.py:
```diff
@@ -70,3 +76,174 @@ def test_loading__invalid__records__threshold():
+    assert json.loads(output[1])['test'] == 'state-2'
+
+
+def test_state__capture__can__be__disabled(capsys):
```
Ah that makes sense, I didn't get that off the bat. I will update!

airhorns commented on target_postgres/state_tracker.py:
```diff
@@ -0,0 +1,78 @@
+import singer.statediff as statediff
+
+
+class TargetError(Exception):
```
Ok, I added that file, definitely feels cleaner

airhorns commented on target_postgres/state_tracker.py:
```diff
@@ -0,0 +1,78 @@
+        all_flushed_watermark = min(self.stream_flush_watermarks.values(), default=0)
+        emittable_state = None
+
+        while len(self.state_queue) > 0 and (force or self.state_queue[0][1] <= all_flushed_watermark):
```
I agree, I think the tuple is unnecessarily opaque! Switched to a dict.

airhorns commented on tests/test_target_tools.py:
```diff
@@ -70,3 +76,174 @@ def test_loading__invalid__records__threshold():
+    assert len(output) == 0
+
+
+def test_state__emits__only__messages__when__all__records__before__have__been__flushed(capsys):
```
💯

airhorns commented:
Comments addressed :)

AlexanderMann commented:
Looks good @airhorns! I'm going to leave this open till end of day to allow others to comment if there's dissent. Pending none, I'll merge!

(PS: like the exceptions module 🥇 )

awm33 commented on target_postgres/stream_tracker.py:
```diff
@@ -0,0 +1,74 @@
+
+class StreamTracker:
+    """
+    Object to track STATE messages coming in from the tap and which streams need to be flushed before they can be safely emitted.
```
nit: Should this be updated to reflect the new class? It does more than tracking state messages.

awm33 commented:
@airhorns I just had one quick comment on updating the docs to that class, but otherwise looks great!

This is an amazing contribution!!!

AlexanderMann commented on target_postgres/stream_tracker.py:
```diff
@@ -0,0 +1,74 @@
+        while len(self.state_queue) > 0 and (force or self.state_queue[0]['watermark'] <= all_flushed_watermark):
+            emittable_state = self.state_queue.popleft()['state']
+
+        if emittable_state:
```
@airhorns just noticed this but shouldn't this emit _all_ states which are safe to emit? ie, this conditional block should really be wrapped up underneath the `while` loop?

AlexanderMann replied:
I'm thinking that the bug here would be if I had a couple states which became safe to emit, this would only emit the latest each time. Not sure if that's expected?

airhorns replied:
My reading of the Singer spec was that this behaviour is ok (and in some circumstances desirable), but, I am totally open to changing this. The reason I think it's ok is that because the state messages are opaque, downstream consumers of them can't really presume to do any merging or munging of the states, so the tap itself has to emit a full state each time. I also think it's desirable to not emit superfluous states because they probably trigger some sort of persistence (it is state after all) downstream, so minimizing the amount of persistence that happens seems ideal. In the instance where a tap emits one state per API call, this is potentially a huge reduction in STATE messages emitted and thus STATE messages processed. 

FWIW this matches `target-snowflake`'s `StateBuffer` doodad here: https://gitlab.com/meltano/target-snowflake/blob/master/target_snowflake/target_snowflake.py#L134-136

airhorns commented:
I added one small commit but big bugfix -- this was only emitting the `value` key of the `STATE` message instead of the whole `STATE` message including the `{"type": "STATE" ...` preamble or any other keys the tap might have added. Oops. I noticed this by actually trying to swap this in for my (rather infantile) infrastructure and figured that out. Updated the tests to expect the whole `STATE` message as well!

airhorns commented:
Thanks folks!
